pageserver/src/page_cache.rs

```diff
@@ -218,18 +219,18 @@ fn init_page_cache(conf: &PageServerConf, timelineid: ZTimelineId) -> PageCache
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
 pub struct CacheKey {
     pub tag: BufferTag,
-    pub lsn: u64,
+    pub lsn: Lsn,
 }
 
 impl CacheKey {
     pub fn pack(&self, buf: &mut BytesMut) {
         self.tag.pack(buf);
-        buf.put_u64(self.lsn);
+        buf.put_u64(self.lsn.0);
```

Comment:
Why is `lsn.0` needed, why not just `lsn`? I thought the `impl From<Lsn> for u64` trait would make that work.

Reply:
The function takes only a u64:
```
fn put_u64(&mut self, n: u64);
```

There are a few ways to get automatic conversions, but none of them happen to be used here:
- Rust always tries to do "autoref" of the `self` argument. This is what allows you to type, e.g. `myvec.len()` instead of `(&my_vec).len()`.
- Some functions have generic parameters that can handle the coercion, e.g. one could write:
```
fn put_u64<T>(&mut self, n: T)
where T: Into<u64>

{
    let n: u64 = n.into();
    // do stuff with n
}
```
- You can do a similar generic type for references using `AsRef` or `AsMut`. You'll see this in the standard library a lot, e.g. `File` has `open<P: AsRef<Path>>(path: P)`, and there's a [long list of `std` types](https://doc.rust-lang.org/std/path/struct.Path.html#impl-AsRef%3CPath%3E) that implement `AsRef<Path`.

The question mark operator (for propagating errors) uses `Into<___>` for whatever your returned error type is. So even if you're not using `anyhow` or `thiserror`, implementing `Into` (or more likely, `From`) for your error types will allow you to use a variety of error types with `?`.

Reply:
I do agree that `buf.put_u64(self.lsn.0)` isn't very friendly.

However, I think that all of this manual byte-pushing code should be replaced with `serde` anyway. As long as we `#[derive(Serialize, Deserialize)]` on the `Lsn` type then it will work exactly the same as u64 with no extra code at all.

Reply:
It's not a big deal, we can live with it. I was just surprised it didn't "just work".

> However, I think that all of this manual byte-pushing code should be replaced with serde anyway. As long as we #[derive(Serialize, Deserialize)] on the Lsn type then it will work exactly the same as u64 with no extra code at all.

That handles this particular case, but I would imagine that this arises in many other places, too. One example off the top of my head: what if you want to store an Lsn in an AtomicU64?

I think this works, too:
```
    buf.put_u64(self.lsn.into());
```

Reply:
I haven't found a good crate that allows `Atomic<T>` where it's implemented as e.g. `AtomicU64` if the size is 64-bits. I made one myself, but never quite finished it. Handling the case where `T` is not `Copy` gets quite nasty, but we don't need that here, so I could probably dig out the simple version (that requires `T = Copy`) if we want it.

Or we could just do a quick `AtomicLsn` wrapper, and wait to see how often we need to cut/paste that code...

Reply:
Another thought: the kind of argument mismatch that `put_u64` hits gets less likely as we get to more complex types (that aren't `Copy`), because they're likely to be passed by reference.

If we have a bunch of functions that take by-reference arguments, e.g.
```
fn handle_mytype(mt: &MyType);
```

Then you'll note that the `mt` argument can be satisfied by any container that `Deref`s to a MyType. For example, `Arc<MyType>` or `MutexGuard<'_, MyType>` both `impl Deref` with `Target=Mytype`, so you can write:
```
handle_mytype(&mt);
```
... and that code will continue to work whether `mt` is a `MyType` or `Arc` or `MutexGuard` (or some other container that we build ourselves).

So your instinct that wrapper types should "just work" is still partially true, at least for the bigger data structures.

Reply:
Will it be better to add pack/unpack methods to Lsn?
So that it is possible to write:
```
self.tag.pack(buf);
self.lsn.pack(buf);
```

Reply:
> Will it be better to add pack/unpack methods to Lsn?

Yes, this can be done automatically with `#[derive(serde::Serialize, serde::Deserialize)]` and then we can use any serde-compatible library we like. The `bincode` crate seems to implement the raw serialization we need in most places (including endian conversions).

I would need a bit more time to look at the code, but the end result is that we should be able to write things like `my_data_structure.serialize(&mut buffer)` without ever needing to type `put_u64` again.

There is one case where I'm not certain about, and that's when the network packet includes little-endian data that can be copied directly into a Rust struct with no validation. I don't think it's possible to generate that code with `serde` right now (see also #6), but if the performance is important to us, I think we can build tools to do that.